### ISSUE TITLE
Calculation of the event size in the payload BUG

### ISSUE THREAD

**Issue by Funkerman1992** [aws/aws-for-fluent-bit#252]
Hi!
In 2.20.0 image I've noticed a set of errors :
`fluent-bit-jpccb fluent-bit [2021/10/02 10:54:33] [ warn] [input] tail.0 paused (mem buf overlimit)`

I've already increased that Mem_buf_limit but still see the `mem buf overlimit`, will increase it one more time, and will keep an eye on it to avoid the OOM situations.
About Chunks I saw next

```
fluent-bit-nplqh fluent-bit time="2021-10-04T15:10:01Z" level=error msg="[cloudwatch 0] InvalidParameterException: Log event too large: 262146 bytes exceeds limit of 262144\n"
fluent-bit-nplqh fluent-bit [2021/10/04 15:10:01] [ warn] [engine] failed to flush chunk '1-1633360196.513460749.flb', retry in 6 seconds: task_id=0, input=tail.0 > output=cloudwatch.0 (out_id=0)
```
Also, saw next errors :
```
fluent-bit-nplqh fluent-bit time="2021-10-04T15:09:57Z" level=error msg="[cloudwatch 0] InvalidParameterException: Log event too large: 262146 bytes exceeds limit of 262144\n"
fluent-bit-nplqh fluent-bit InvalidParameterException: Log event too large: 262146 bytes exceeds limit of 262144
```

Maybe I need to increase more options of the input plugin? I'm already changed from defaults

```
        Buffer_Chunk_Size   64k
        Buffer_Max_Size     128k
        Mem_Buf_Limit       150MB
        Skip_Long_Lines     Off
```

```
fluent-bit-nplqh fluent-bit [2021/10/04 15:09:56] [ warn] [engine] failed to flush chunk '1-1633360191.515079981.flb', retry in 11 seconds: task_id=1, input=tail.0 > output=cloudwatch.0 (out_id=0)
fluent-bit-nplqh fluent-bit [2021/10/04 15:09:56] [ warn] [engine] failed to flush chunk '1-1633360192.222293300.flb', retry in 10 seconds: task_id=3, input=tail.0 > output=cloudwatch.0 (out_id=0)
```
`I1005 15:00:08.379771   72624 request.go:665] Waited for 10.886164692s due to client-side throttling, not priority and fairness, request: GET:https://90DA65ABC7C608CEDE2537EF0548C42D.gr7.us-east-1.eks.amazonaws.com/api/v1/namespaces/amazon-cloudwatch/pods/fluent-bit-lg8d9/log?container=fluent-bit&follow=true&sinceSeconds=1`

```
fluent-bit-pkc2v fluent-bit [2021/10/05 12:01:26] [ info] [input] tail.0 resume (mem buf overlimit)
fluent-bit-fsfv7 fluent-bit [2021/10/05 12:01:26] [ info] [input:tail:tail.0] inode=130056895 link_rotated: /var/log/containers/account-rank-predict-1431-2-bm68x_analytics_account-rank-predict-1431-2-1af426256406d9c2f9c93f66e9a5d9785b0a75c343c754c8c6e9173da19da4e9.log
fluent-bit-fsfv7 fluent-bit [2021/10/05 12:01:26] [ info] [input:tail:tail.0] inode=480268471 handle rotation(): /var/log/containers/account-rank-predict-1431-2-bm68x_analytics_account-rank-predict-1431-2-1af426256406d9c2f9c93f66e9a5d9785b0a75c343c754c8c6e9173da19da4e9.log => /var/lib/docker/containers/1af426256406d9c2f9c93f66e9a5d9785b0a75c343c754c8c6e9173da19da4e9/1af426256406d9c2f9c93f66e9a5d9785b0a75c343c754c8c6e9173da19da4e9-json.log (deleted)
```

```
fluent-bit-kr79n fluent-bit InvalidParameterException: Log event too large: 262145 bytes exceeds limit of 262144
I1005 15:01:29.775838   72624 request.go:665] Waited for 1m31.624727538s due to client-side throttling, not priority and fairness, request: GET:https://90DA65ABC7C608CEDE2537EF0548C42D.gr7.us-east-1.eks.amazonaws.com/api/v1/namespaces/amazon-cloudwatch/pods/fluent-bit-vft45/log?container=fluent-bit&follow=true&sinceSeconds=1
fluent-bit-kr79n fluent-bit [2021/10/05 12:01:29] [ warn] [engine] failed to flush chunk '1-1633435285.379613366.flb', retry in 7 seconds: task_id=0, input=tail.0 > output=cloudwatch.0 (out_id=0)
fluent-bit-572fc fluent-bit [2021/10/05 12:01:30] [ info] [input:tail:tail.0] inode=37749341 handle rotation(): /var/log/containers/chi-ch-1-ch-1-0-0-0_ch-1_ch-50308ca52f3f980140c1a734bf66130ba59fe9326bfd35c57c6829b4381c57eb.log => /var/lib/docker/containers/50308ca52f3f980140c1a734bf66130ba59fe9326bfd35c57c6829b4381c57eb/50308ca52f3f980140c1a734bf66130ba59fe9326bfd35c57c6829b4381c57eb-json.log.1
```
No more error messages in my fluentbit application log.

My current FluentBit config config :
```
apiVersion: v1
kind: ConfigMap
metadata:
  name: fluent-bit-config
  namespace: amazon-cloudwatch
  labels:
    k8s-app: fluent-bit
data:
  fluent-bit.conf: |
    [SERVICE]
        Flush                     5
        Log_Level                 info
        Daemon                    off
        Parsers_File              parsers.conf
        HTTP_Server               ${HTTP_SERVER}
        HTTP_Listen               0.0.0.0
        HTTP_Port                 ${HTTP_PORT}
        storage.path              /var/fluent-bit/state/flb-storage/
        storage.sync              normal
        storage.checksum          off
        storage.backlog.mem_limit 5M

    @INCLUDE application-log.conf
    @INCLUDE dataplane-log.conf
    @INCLUDE host-log.conf

  application-log.conf: |
    [INPUT]
        Name                tail
        Tag                 application.*
        Exclude_Path        /var/log/containers/cloudwatch-agent*, /var/log/containers/fluent-bit*, /var/log/containers/aws-node*, /var/log/containers/kube-proxy*
        Path                /var/log/containers/*.log
        Docker_Mode         On
        Docker_Mode_Flush   5
        Docker_Mode_Parser  container_firstline
        Parser              docker
        DB                  /var/fluent-bit/state/flb_container.db
        Buffer_Chunk_Size   64k
        Buffer_Max_Size     128k
        Mem_Buf_Limit       150MB
        Skip_Long_Lines     Off
        Refresh_Interval    10
        Rotate_Wait         30
        storage.type        filesystem
        Read_from_Head      ${READ_FROM_HEAD}

    [FILTER]
        Name                kubernetes
        Match               application.*
        Kube_URL            https://kubernetes.default.svc:443
        Kube_Tag_Prefix     application.var.log.containers.
        Merge_Log           On
        Merge_Log_Key       log_processed
        K8S-Logging.Parser  On
        K8S-Logging.Exclude Off
        Labels              On
        Annotations         Off

    [OUTPUT]
        Name                cloudwatch
        Match               application.*
        region              ${AWS_REGION}
        default_log_group_name FluentBit-Application
        log_group_name      /aws/containerinsights/${CLUSTER_NAME}/$(kubernetes['namespace_name'])/$(kubernetes['labels']['app'])
        log_stream_prefix   ${HOST_NAME}-
        auto_create_group   true
        extra_user_agent    container-insights

  dataplane-log.conf: |
    [INPUT]
        Name                systemd
        Tag                 dataplane.systemd.*
        Systemd_Filter      _SYSTEMD_UNIT=docker.service
        DB                  /var/fluent-bit/state/systemd.db
        Path                /var/log/journal
        Read_From_Tail      ${READ_FROM_TAIL}

    [INPUT]
        Name                tail
        Tag                 dataplane.tail.*
        Path                /var/log/containers/aws-node*, /var/log/containers/kube-proxy*
        Docker_Mode         On
        Docker_Mode_Flush   5
        Docker_Mode_Parser  container_firstline
        Parser              docker
        DB                  /var/fluent-bit/state/flb_dataplane_tail.db
        Mem_Buf_Limit       50MB
        Skip_Long_Lines     On
        Refresh_Interval    10
        Rotate_Wait         30
        storage.type        filesystem
        Read_from_Head      ${READ_FROM_HEAD}

    [FILTER]
        Name                modify
        Match               dataplane.systemd.*
        Rename              _HOSTNAME                   hostname
        Rename              _SYSTEMD_UNIT               systemd_unit
        Rename              MESSAGE                     message
        Remove_regex        ^((?!hostname|systemd_unit|message).)*$

    [FILTER]
        Name                aws
        Match               dataplane.*
        imds_version        v1

    [OUTPUT]
        Name                cloudwatch
        Match               dataplane.*
        region              ${AWS_REGION}
        log_group_name      /aws/containerinsights/${CLUSTER_NAME}-fluentbit/dataplane
        log_stream_prefix   ${HOST_NAME}-
        auto_create_group   true
        extra_user_agent    container-insights

  host-log.conf: |
    [INPUT]
        Name                tail
        Tag                 host.dmesg
        Path                /var/log/dmesg
        Parser              syslog
        DB                  /var/fluent-bit/state/flb_dmesg.db
        Mem_Buf_Limit       5MB
        Skip_Long_Lines     On
        Refresh_Interval    10
        Read_from_Head      ${READ_FROM_HEAD}

    [INPUT]
        Name                tail
        Tag                 host.messages
        Path                /var/log/messages
        Parser              syslog
        DB                  /var/fluent-bit/state/flb_messages.db
        Mem_Buf_Limit       5MB
        Skip_Long_Lines     On
        Refresh_Interval    10
        Read_from_Head      ${READ_FROM_HEAD}

    [INPUT]
        Name                tail
        Tag                 host.secure
        Path                /var/log/secure
        Parser              syslog
        DB                  /var/fluent-bit/state/flb_secure.db
        Mem_Buf_Limit       5MB
        Skip_Long_Lines     On
        Refresh_Interval    10
        Read_from_Head      ${READ_FROM_HEAD}

    [FILTER]
        Name                aws
        Match               host.*
        imds_version        v1

    [OUTPUT]
        Name                cloudwatch
        Match               host.*
        region              ${AWS_REGION}
        log_group_name      ${CLUSTER_NAME}-fluentbit/host
        log_stream_prefix   ${HOST_NAME}.
        auto_create_group   true
        extra_user_agent    container-insights

  parsers.conf: |
    [PARSER]
        Name                docker
        Format              json
        Time_Key            time
        Time_Format         %Y-%m-%dT%H:%M:%S.%LZ

    [PARSER]
        Name                syslog
        Format              regex
        Regex               ^(?<time>[^ ]* {1,2}[^ ]* [^ ]*) (?<host>[^ ]*) (?<ident>[a-zA-Z0-9_\/\.\-]*)(?:\[(?<pid>[0-9]+)\])?(?:[^\:]*\:)? *(?<message>.*)$
        Time_Key            time
        Time_Format         %b %d %H:%M:%S

    [PARSER]
        Name                container_firstline
        Format              regex
        Regex               (?<tag>[^.]+)?\.?(?<pod_name>[a-z0-9](?:[-a-z0-9]*[a-z0-9])?(?:\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*)_(?<namespace_name>[^_]+)_(?<container_name>.+)-(?<docker_id>[a-z0-9]{64})\.log$
        Time_Key            time
        Time_Format         %Y-%m-%dT%H:%M:%S.%LZ

    [PARSER]
        Name                cwagent_firstline
        Format              regex
        Regex               (?<log>(?<="log":")\d{4}[\/-]\d{1,2}[\/-]\d{1,2}[ T]\d{2}:\d{2}:\d{2}(?!\.).*?)(?<!\\)".*(?<stream>(?<="stream":").*?)".*(?<time>\d{4}-\d{1,2}-\d{1,2}T\d{2}:\d{2}:\d{2}\.\w*).*(?=})
        Time_Key            time
        Time_Format         %Y-%m-%dT%H:%M:%S.%LZ
```

**Comment by matthewfala** [comment]
Hi @Funkerman1992  👋 


> "fluent-bit-kr79n fluent-bit InvalidParameterException: Log event too large: 262145 bytes exceeds limit of 262144" 

We determined that your second issue likely stems from our truncation system, which removes excess bytes off the log to satisfy cloudwatch's maximumBytesPerEvent. This is a problem if the the bytes at the truncation point are a multibyte unicode character. The truncation creates invalid unicode sequence. Later when the resulting truncated log containing the unicode sequence is converted to a string from byte slice, the sequence expands to escaped hex format (\x{some-hex}{some-hex}). This gets interpreted as 4 unique characters when sent to cloudwatch. The addition of around 3 bytes after truncation sparks a log event too large exception.

Here is the related code

```
	// append newline
	data = append(data, []byte("\n")...)

	if (len(data) + perEventBytes) > maximumBytesPerEvent {
		logrus.Warnf("[cloudwatch %d] Found record with %d bytes, truncating to 256KB, logGroup=%s, stream=%s\n",
			output.PluginInstanceID, len(data)+perEventBytes, e.group, e.stream)
		data = data[:(maximumBytesPerEvent - len(truncatedSuffix) - perEventBytes)]
		data = append(data, []byte(truncatedSuffix)...)
	}

	return data, nil
```
and
```
// Takes the byte slice and returns a string
// Also removes leading and trailing whitespace
func logString(record []byte) string {
	return strings.TrimSpace(string(record))
}
```

The attached log, was generated to assess this error. If sent to fluent bit, fluent bit will throw invalidParameterException.

### Documentation about the log payload:
> A special, 4 byte, unicode character starts at byte 262104 in the log. The log is at least 262117 bytes long.

```
(Send log payload to fluent bit): ...mnwqhk𒁈zrgchimqigtm"} -> Fluent Bit
(Fluent Bit) route to Go Cloudwatch plugin
# Convert ...mnwqhk𒁈zrgchimqigtm\"}\n"
#    to byte array ...,109,110,119,113,104,107,240,146,129,136,122,114,103,99,104,105,109,113,105,103,116,109,34,125,10]
#      - Note that 240, 146, 129, and 134 are all out of ASCII range [0-127]
# Truncate log, and discard bytes mid special character.
#			...,109,110,119,113,104,107,240, |-> Truncate rest

# Convert back to string with: string([]Byte)
#     "mnwqhk\xf0"
#        - Note: Go automatically detects undetected characterset, and describes it as escaped hex
#                   JSON does not have escaped hex, and thus, \xf0 is 4 unique characters
#                   1 character becomes 4 characters, so 2~3 characters have been added past maximum
# Cloudwatch is sent the following payload:
#    ...mnwqhk\xf0[Truncated...]\",\"timestamp\":1633655429180}],\"logGroupName\":\"fluent_group\",\"logStreamName\":\"testapp.log\",\"sequenceToken\":\"reda\"}\r\n"

(Cloud Watch) receives request and returns:
#     The message "262146 bytes exceeds limit of 262144" from API
#     262143 + 3 characters = 262146 characters

#      We truncate to 262143 characters which is 1 character less than we need to. why? Not sure...
```

### Log Payload
https://tiny.amazon.com/ob1zb4oq/paste

**Comment by matthewfala** [comment]
@Funkerman1992 
I have published a PR which I believe will resolve the "Log event too large" issue. Are you able to build and test the plugin? If not I may have time tomorrow to build for you. How are you running fluent bit?

**Comment by matthewfala** [comment]
I merged the patch in. It should be included in our next release, which will be either aws-for-fluent-bit v2.21.2 or v2.22.0

https://github.com/aws/amazon-cloudwatch-logs-for-fluent-bit/pull/208

**Comment by zhonghui12** [comment]
The solution has been released in aws-for-fluent-bit v2.21.2, please try: https://github.com/aws/aws-for-fluent-bit/releases/tag/v2.21.2.
